src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -3349,6 +3451,331 @@ public <T> T invokeAny(Collection<? extends Callable<T>> tasks,
             .invokeAny(tasks, this, true, unit.toNanos(timeout));
     }
 
+    // Support for delayed tasks
+
+    /**
+     *  Creates and starts DelayScheduler
+     */
+    private DelayScheduler startDelayScheduler() {
+        DelayScheduler ds;
+        if ((ds = delayScheduler) == null) {
+            boolean start = false;
+            String name = poolName + "-delayScheduler";
+            if (workerNamePrefix == null)
+                asyncCommonPool();  // override common parallelism zero
+            lockRunState();
+            try {
+                if ((ds = delayScheduler) == null) {
+                    ds = delayScheduler = new DelayScheduler(this, name);
+                    start = true;
+                }
+            } finally {
+                unlockRunState();
+            }
+            if (start) { // start outside of lock
+                // exceptions on start passed to (external) callers
+                SharedThreadContainer ctr;
+                if ((ctr = container) != null)
+                    ctr.start(ds);
+                else
+                    ds.start();
+            }
+        }
+        return ds;
+    }
+
+    /**
+     * Callback upon starting DelayScheduler
+     */
+    final void onDelaySchedulerStart() {
+        WorkQueue q;           // set up default submission queue
+        if ((q = submissionQueue(0, false)) != null)
+            q.unlockPhase();
+    }
+
+    /**
+     * Arranges execution of a ready task from DelayScheduler, or
+     * cancels it on error
+     */
+    final void executeReadyScheduledTask(ScheduledForkJoinTask<?> task) {
+        if (task != null) {
+            WorkQueue q;
+            boolean cancel = false;
+            try {
+                if ((q = externalSubmissionQueue(false)) == null)
+                    cancel = true; // terminating
+                else
+                    q.push(task, this, false);
+            } catch(Error | RuntimeException ex) {
+                cancel = true;     // OOME or VM error
+            }
+            if (cancel)
+                task.trySetCancelled();
+        }
+    }
+
+    /**
+     * Arranges delayed execution of a ScheduledForkJoinTask via the
+     * DelayScheduler, creating and starting it if necessary.
+     * @return the task
+     */
+    final <T> ScheduledForkJoinTask<T> scheduleDelayedTask(ScheduledForkJoinTask<T> task) {
+        DelayScheduler ds;
+        if (((ds = delayScheduler) == null &&
+             (ds = startDelayScheduler()) == null) ||
+            task == null || (runState & SHUTDOWN) != 0L)
```

Comment:
Might not make sense to start the delay scheduler if the task is null.

Reply:
There's no longer need or value of that null check (there was in some trial version), so removed.